File: solutions/JavaScript/word_break_dp_set.js
/**
 * Determine if string can be segmented using words from dictionary.
 *
 * Time Complexity: O(n²)
 * Space Complexity: O(n + m)
 */
function wordBreakDpSet(s, wordDict) {
    const wordSet = new Set(wordDict);
    const dp = new Array(s.length + 1).fill(false);
    dp[0] = true;

    for (let i = 1; i <= s.length; i++) {
        for (let j = 0; j < i; j++) {
            if (dp[j] && wordSet.has(s.substring(j, i))) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[s.length];
}

console.log(wordBreakDpSet("leetcode", ["leet", "code"]));  // true
console.log(wordBreakDpSet("applepenapple", ["apple", "pen"]));  // true
console.log(wordBreakDpSet("catsandog", ["cat", "cats", "and", "sand", "dog"]));  // false

File: solutions/Java/WordBreakDpSet.java
import java.util.*;

/**
 * Determine if string can be segmented using words from dictionary.
 *
 * Time Complexity: O(n²)
 * Space Complexity: O(n + m)
 */
public class WordBreakDpSet {
    public static boolean wordBreakDpSet(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<>(wordDict);
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;

        for (int i = 1; i <= s.length(); i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && wordSet.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[s.length()];
    }

    public static void main(String[] args) {
        System.out.println(wordBreakDpSet("leetcode", Arrays.asList("leet", "code")));  // true
        System.out.println(wordBreakDpSet("applepenapple", Arrays.asList("apple", "pen")));  // true
        System.out.println(wordBreakDpSet("catsandog", Arrays.asList("cat", "cats", "and", "sand", "dog")));  // false
    }
}
